Adding information about debugging

sishida opened the pull request:
Signed-off-by: Sarah Ishida <sishida@us.ibm.com>

PR for issue https://github.com/eclipse/codewind/issues/1990

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,33 @@ type: document
+- When you debug initialization code, wait for the project state to change to **Debugging** or for the debugger to stop at a breakpoint.
+- Normally you can reload your application multiple times to isolate a problem. However, when you debug initialization code, restart your project in debug mode to stop in the code again.
+
+## Debugging Node.js projects
```
Is this a shared document with VSCode and IntelliJ? This part does not apply to VSCode. Not sure about IntelliJ.

sishida replied:
John specified that it doesn't apply to IntelliJ. So looks like this Node.js section applies to Eclipse only. I will update the docs accordingly. :) Thanks!

sishida commented:
@eharris369 Thanks for your feedback. I have added a new commit to this PR. 👍

sishida commented:
Hi, @tetchel, I have added info about remote debugging from your comment in the issue https://github.com/eclipse/codewind/issues/1990#issuecomment-621347424 into the latest commit to this PR.

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -18,6 +18,10 @@ Complete the following steps to debug in Codewind applications:
 4. While the project is still in Debugging mode, if the debug process is detached, right-click your project and select **Attach Debugger** from the menu. This action reattaches the debugger to the debug process, and you can continue with debugging your application.
 5. To switch back to run mode, right-click your project and select **Restart in Run Mode** from the menu.
 
+## Debugging remotely
```
Did you see the comments I added to https://github.com/eclipse/codewind/issues/2651 including some Eclipse specific information?

sishida replied:
Oh--I'll take a look at that! Thank you!

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,41 @@ type: document
+Remote debugging is the same as local debugging but with the following exceptions:
+- Install `kubectl` or `oc` and make sure you are logged in to your cluster.
+- After the project restarts in debug mode, Codewind automatically sets up a port forward from the application's debug port to a port on your local machine.
+   - In Eclipse, this port forward process shows in the Debug view. If you end the port forward, Codewind ends your debug session.
```
Codewind does not end the debug session. The debug session just dies because it looses the connection.

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,41 @@ type: document
+- Install `kubectl` or `oc` and make sure you are logged in to your cluster.
+- After the project restarts in debug mode, Codewind automatically sets up a port forward from the application's debug port to a port on your local machine.
+   - In Eclipse, this port forward process shows in the Debug view. If you end the port forward, Codewind ends your debug session.
+   - For Java projects, if you detach the debugger or restart the application in run mode, both the debug session and the port forward are removed automatically.
```
Say: both the debug session and the port forward are terminated automatically

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,41 @@ type: document
+5. To switch back to run mode, right-click your project and select **Restart in Run Mode** from the menu.
+
+## Debugging remotely
+Remote debugging is the same as local debugging but with the following exceptions:
```
The 'but' seems redundant.

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,41 @@ type: document
+- Install `kubectl` or `oc` and make sure you are logged in to your cluster.
+- After the project restarts in debug mode, Codewind automatically sets up a port forward from the application's debug port to a port on your local machine.
+   - In Eclipse, this port forward process shows in the Debug view. If you end the port forward, Codewind ends your debug session.
+   - For Java projects, if you detach the debugger or restart the application in run mode, both the debug session and the port forward are removed automatically.
```
This may be true for Node projects as well in VSCode. Not sure if this whole section is Eclipse specific or not.

sishida replied:
@tetchel Hi, Tim, Do you happen to know if this "Debugging remotely" section also applies to VS Code? Thanks!

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,41 @@ type: document
+- After the project restarts in debug mode, Codewind automatically sets up a port forward from the application's debug port to a port on your local machine.
+   - In Eclipse, this port forward process shows in the Debug view. If you end the port forward, Codewind ends your debug session.
+   - For Java projects, if you detach the debugger or restart the application in run mode, both the debug session and the port forward are removed automatically.
+   - For Node.js applications where you use a Chromium-based browser for debugging, manually end the associated port forwarding process in the Debug view of Eclipse if you end the debug session in your browser.
```
This should be called out as Eclipse specific at the start, unless this whole section is supposed to be Eclipse specific.

tetchel commented:
This feature has been pulled from 0.12 so I'm moving the PR to draft

tetchel commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,41 @@ type: document
+## Debugging remotely
+Remote debugging is the same as local debugging with the following exceptions:
+- Install `kubectl` or `oc` and make sure you are logged in to your cluster.
+- In Eclipse, after the project restarts in debug mode, Codewind automatically sets up a port forward from the application's debug port to a port on your local machine.
```
this is true in vs code too

tetchel replied:
we could say something like:

```
In VS Code, the port forward is run as a VS Code task, which appears in the *Terminal* view. 
```

tetchel replied:
I don't know where it appears in Eclipse.

tetchel commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,41 @@ type: document
+- In Eclipse, after the project restarts in debug mode, Codewind automatically sets up a port forward from the application's debug port to a port on your local machine.
+   - If you end the port forward and have debug sessions that depend on that port forward, those debug sessions also end.
+   - For Java projects, if you detach the debugger or restart the application in run mode, both the debug session and the port forward are terminated automatically.
+   - For Node.js applications where you use a Chromium-based browser for debugging, manually end the associated port forwarding process in the Debug view if you end the debug session in your browser.
```
this is eclipse only

tetchel commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,41 @@ type: document
+- Install `kubectl` or `oc` and make sure you are logged in to your cluster.
+- In Eclipse, after the project restarts in debug mode, Codewind automatically sets up a port forward from the application's debug port to a port on your local machine.
+   - If you end the port forward and have debug sessions that depend on that port forward, those debug sessions also end.
+   - For Java projects, if you detach the debugger or restart the application in run mode, both the debug session and the port forward are terminated automatically.
```
this is not quite true for vs code.
 If you close the debug session in vs code, the port-forward stays active. 
If you close the port forward, the debug session terminates because the debug port has closed.

tetchel commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,41 @@ type: document
+   - For Java projects, if you detach the debugger or restart the application in run mode, both the debug session and the port forward are terminated automatically.
+   - For Node.js applications where you use a Chromium-based browser for debugging, manually end the associated port forwarding process in the Debug view if you end the debug session in your browser.
+
+## Using Java hot code replace while debugging
```
This and the 'debugging initialization' section both apply to both IDEs.

eharris369 commented:
Can the information about debugging initialization for Java and debugging Node.js projects in Eclipse be moved to another PR since this should go in for 0.12.0?

eharris369 commented:
I created this PR for 0.12.0 to restore the missing debugging information related to hot code replace, debugging initialization and debugging Node.js applications with Eclipse: https://github.com/eclipse/codewind-docs/pull/574

sishida commented:
Hi, @eharris369, Tim specified that "In VS Code, the port forward is run as a VS Code task, which appears in the Terminal view." Could you please let me know the equivalent in Eclipse? Thank you! 😃

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,45 @@ type: document
+- Install `kubectl` or `oc` and make sure you are logged in to your cluster.
+- After the project restarts in debug mode, Codewind automatically sets up a port forward from the application's debug port to a port on your local machine.
+   - In VS Code, the port forward is run as a VS Code task, which appears in the Terminal view.
+   - <Do we want to specify where this appears in Eclipse?>
```
A launch is created for the port forward that appears in the Debug view.

eharris369 commented:
@sishida A launch is created for the port forward that appears in the Debug view.

sishida commented:
**Note:** We have already merged this PR https://github.com/eclipse/codewind-docs/pull/574 to be included with the 0.12.0 release. If merge conflicts occur when we try to merge PR 560, we might have to remove the info from 560 that was already merged in 574.

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,45 @@ type: document
+   - In Eclipse, for Node.js applications where you use a Chromium-based browser for debugging, manually end the associated port forwarding process in the Debug view if you end the debug session in your browser.
+   - In Eclipse, for Java projects, if you detach the debugger or restart the application in run mode, both the debug session and the port forward are terminated automatically.
+   - In VS Code, for Java projects, if you close the debug session, the port forward stays active.
+If you close the port forward, the debug session terminates because the debug port has closed.
```
This is covered above in this point:
- If you end the port forward and have debug sessions that depend on that port forward, those debug sessions also end.

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,43 @@ type: document
+   - In VS Code, the port forward is run as a VS Code task, which appears in the Terminal view.
+   - In Eclipse, for Node.js applications where you use a Chromium-based browser for debugging, manually end the associated port forwarding process in the Debug view if you end the debug session in your browser.
+   - In Eclipse, for Java projects, if you detach the debugger or restart the application in run mode, both the debug session and the port forward are terminated automatically.
+   - In VS Code, for Java projects, if you close the debug session, the port forward stays active. If you close the port forward, the debug session terminates because the debug port has closed.
```
The part about how terminating the port forward will terminate the debug session is true for Eclipse as well.  Maybe:
- In VS Code, for Java projects, if you close the debug session, the port forward stays active.
- In VS Code and Eclipse if you close the port forward, the debug session terminates because the debug port has closed.
@tetchel The first point says it relates to Java projects only, does the port forward get cleaned up automatically for Node projects?

tetchel replied:
in vs code, ending the debug session does not ever end the port forward. I figured we might as well leave it in case they want to re-attach the debugger. The user can close the terminal by hand if they wish to end the port forward.

eharris369 replied:
@sishida Based on Tim's comment above this should be changed to:
   -  In VS Code if you close the debug session, the port forward stays active.
   -  In VS Code and Eclipse if you close the port forward, the debug session terminates because the debug port has closed.

tetchel replied:
I think we said this elsewhere but for clarity we could say

In VS Code if you close the debug session, the port forward task stays active in the Terminal view.

sishida commented:
@eharris369 @tetchel Thanks for the feedback! I have added it in the latest commit.

tetchel commented on docs/_documentations/debugging.md:
```diff
@@ -10,10 +10,44 @@ type: document
+   - In Eclipse, for Node.js applications where you use a Chromium-based browser for debugging, manually end the associated port forwarding process in the Debug view if you end the debug session in your browser.
+   - In Eclipse, for Java projects, if you detach the debugger or restart the application in run mode, both the debug session and the port forward are terminated automatically.
+   - In VS Code, if you close the debug session, the port forward task stays active in the Terminal view.
+   - In Eclipse and VS Code, if you close the port forward, the debug session terminates because the debug port has closed.
```
It's not really correct to say the port 'has closed'. Maybe we could say 'the debug port is no longer accessible [from the local machine]'.

sishida replied:
@tetchel Thanks for the clarification! Update made in latest commit.

sishida commented:
**Note:** This PR is **not** to be included with 0.12.0. However, it has developer approval and is ready to go **after** 0.12.0.

sishida commented:
Hi, @eharris369, Could you please look at the latest commit (https://github.com/eclipse/codewind-docs/pull/560/commits/eea4d88304b3e7432830e5357557e8d00a6c85bc) and check if it addresses your comment about debugging remote Java projects (https://github.com/eclipse/codewind/issues/2651)? Thanks!

eharris369 commented on docs/_documentations/debugging.md:
```diff
@@ -22,8 +22,10 @@ Complete the following steps to debug in Codewind applications:
 - After the project restarts in debug mode, Codewind automatically sets up a port forward from the application's debug port to a port on your local machine.
-   - In Eclipse, a launch is created for the port forward that appears in the Debug view.
+**Starting a debug session**
+   - In Eclipse Node.js projects, a launch is created for the port forward that appears in the Debug view. In Eclipse Java projects, the port forward is added as a process to the debug launch instead of a separate port forward launch.
```
How about:
In Eclipse, for Node.js projects, a launch is created for the port forward that appears in the Debug view. For Java projects, the port forward is added as a process to the debug launch.

sishida commented:
@eharris369 Thanks for the feedback! I've added it in the latest commit.
